### ISSUE TITLE
ccls does not work with non-root compile_commands.json

### ISSUE THREAD

**Issue by QuLogic** [dense-analysis/ale#2621]
## Information

**VIM version**

NVIM v0.3.6
Build type: RelWithDebInfo

Operating System: Fedora 30

## What went wrong

`ccls` linting does not work correctly when `compile_commands.json` is in a build subdirectory. Because of 143af2b9fd78dbfd5ee8ae727a522dd2dcc79b61, the project _root_ is correctly found, but the build dir is not told to `ccls`. This results in several linting errors because it doesn't have the right `#define`s or include paths. A workaround is to symlink the file to the project root.

`ccls` provides [an initialization option](https://github.com/MaskRay/ccls/wiki/Project-Setup#compile_commandsjson) called `compilationDatabaseDirectory` that could be set to the build dir that ale finds. Note, it does _not_ work to simply set `let g:ale_c_ccls_init_options = {'compilationDatabaseDirectory': 'build'}` (I assume because either it needs to be an absolute path or the working directory is not the project root.)

This may also affect `clangd` which has a `-compile-commands-dir=` option that could be used similarly, but I have not tested it.

## Reproducing the bug

1. Start with some C project with an out-of-source build, using a `build` subdirectory in the project root, i.e., have source files in `$PROJECT_ROOT` and `compile_commands.json` in `$PROJECT_ROOT/build/compile_commands.json`.
2. Symlink it to the project root: `ln -s $PROJECT_ROOT/build/compile_commands.json $PROJECT_ROOT/compile_commands.json`
3. `ccls` linting works.
4. Delete the symlink.
5. `ccls` linting stops working.

### :ALEInfo

```
 Current Filetype: c
Available Linters: ['ccls', 'clang', 'clangd', 'clangtidy', 'cppcheck', 'cquery', 'flawfinder', 'gcc']
  Enabled Linters: ['ccls', 'clang', 'clangd', 'clangtidy', 'cppcheck', 'cquery', 'flawfinder', 'gcc']
 Suggested Fixers: 
  'clang-format' - Fix C/C++ and cuda files with clang-format.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_c_ccls_executable = 'ccls'
let g:ale_c_ccls_init_options = {}
let g:ale_c_clang_executable = 'clang'
let g:ale_c_clang_options = '-std=c11 -Wall'
let g:ale_c_clangd_executable = 'clangd'
let g:ale_c_clangd_options = ''
let g:ale_c_clangtidy_checks = []
let g:ale_c_clangtidy_executable = 'clang-tidy'
let g:ale_c_clangtidy_extra_options = ''
let g:ale_c_clangtidy_options = ''
let g:ale_c_cppcheck_executable = 'cppcheck'
let g:ale_c_cppcheck_options = '--enable=style'
let g:ale_c_cquery_cache_directory = '/home/elliott/.cache/cquery'
let g:ale_c_cquery_executable = 'cquery'
let g:ale_c_flawfinder_error_severity = 6
let g:ale_c_flawfinder_executable = 'flawfinder'
let g:ale_c_flawfinder_minlevel = 1
let g:ale_c_flawfinder_options = ''
let g:ale_c_gcc_executable = 'gcc'
let g:ale_c_gcc_options = '-std=c11 -Wall'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_history_enabled = 0
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {'spec': ['rpmlint'], 'python': ['flake8']}
let g:ale_linters_explicit = 0
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = v:null
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = {'\.min.js$': {'ale_enabled': 0}}
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:
```

**Comment by mrgaturus** [comment]
using ̣`let g:ale_c_clangd_options = '-compile-commands-dir=build'` (where `build` is the directory of compile_commands.json) on clangd workarounds the problem

**Comment by rustushki** [comment]
I can confirm that clangd is affected by this problem as well, and also that the workaround by @mrgaturus also works.

**Comment by w0rp** [comment]
I've created a pull request here which might fix this for both linters. Let me know what you think. #2713

**Comment by liskin** [comment]
I ran into this yesterday myself and, assuming it isn't fixed yet, created my own pull request here: #3123. Turns out there's #2713 already, but without tests and with conflicts, so I guess we can just use mine. :-)

**Comment by w0rp** [comment]
I've merged that pull request now. Thanks @liskin. :+1:
